### ISSUE TITLE
Async updates updates old states.

### ISSUE THREAD

**Issue by zaceno** [zaceno/hyperapp-partial#3]
There is a problem when using thunks for async updates. If you have an action which fetches some data, and then updates the state with that data, the new state will be merged with the state as it was before the fetch. And *any updates which happened in between will be reset*.

Working on a fix.

**Comment by jbucaran** [comment]
@zaceno 🤔 How could you fix this? The thunk is "closed" on the state as it was when it was returned. The only thing I can imagine right now is to create a "getter" that returns the state, so you can get a fresh copy of the state inside the callback when you use the `update` function.

**Comment by zaceno** [comment]
@jbucaran I don't mean it's a bug in hyperapp. Rather, it's a bug in hyperapp-partial, in the sense that I don't deliver on the promise that partials should behave like regular apps, in terms of the state they have access to.

And you're right, the only way to solve the problem is to make a getter to access the global state when I call my scoped updates.

**Comment by RikuVan** [comment]
oh given this bug, am I going to have problems with something like this in a partial?
```
actions: {
      dispatch: ( s, a, payload) => {
        a.dismiss()
        return {active: payload}
      },

      dismiss: s => update => {
        delay(s.delay, () => update({active: {}}))
      },
}
```

**Comment by zaceno** [comment]
@RikuVan Yes, you might. The what you may find is that other updates to the state (within the same partial), which occur *during* the delay, seem to be "reset" once the delay finishes.

**Comment by zaceno** [comment]
I am very close to fixing this bug btw. I have a basic solution, which works and most tests are passing. 

But it did require a bit of a refactor, so some tests are still not passing -- 99% likely due to my moving stuff around, and not indicative of a problem with the fix. 

As soon as the tests are working again, I'll push 0.4.1 with the fix.

**Comment by jbucaran** [comment]
@zaceno Sorry for the noise, but is this an issue with hyperapp-partial?

Because given @RikuVan's code I don't see any problem, unless he uses the value of the state in that closure to determine the next state.

**Comment by zaceno** [comment]
@jbucaran -- yes, it's a particular problem of hyperapp-partial. Because at the app level you can do

```
update({foo: 'newfoo'})
```

and it will merge foo with the current state: 

```
{foo: 'oldfoo', bar: 'bar'} --> {foo: 'newfoo', bar: 'bar'}
```

However, hyperapp-partial tries to emulate this shallow-merge-behavior, but it's actually at a deeper point in the state, which means I must "deep-upsert" the new state-prop passed to `update`, into a full state object, before passing that to hyperapp's own `update` function. And the state object I upsert into is ... you guessed it: the old state passed to the action.

**Comment by zaceno** [comment]
Again, I'm not far from a fix -- but the world seems conspiring against me. I haven't had time to work on it yet.

**Comment by zaceno** [comment]
@jbucaran hyperapp#347 will make my fix *much* simpler :)

**Comment by jbucaran** [comment]
@zaceno It's merged! Will you need anything else

**Comment by zaceno** [comment]
@jbucaran nope, just need time ;) But thanks!

**Comment by zaceno** [comment]
Finally! I got a solid block of hours to myself to work on this (long train ride! 🚂 ). So the fix is merged to master: https://github.com/zaceno/hyperapp-partial/commit/c12e25a4087addd91a930040eaa1c4d30b6080e3

The solution is built on https://github.com/hyperapp/hyperapp/pull/347  (reducer-thunks), and simultaneously adds support for reducer-thunks in partials.

Reducer-thunks are not currently on hyperapp's master, but not yet in any release. This means you can't use this fix yet, unless you're also building hyperapp yourself from master.

As soon as there's a release of hyperapp with reducer thunks, I'll push a hyperapp-partial release containing the fix.

**Comment by zaceno** [comment]
Ok so now hyperapp 0.12.1 is out, and I published hyperapp-partial 0.5.0 -- so I think this issue is handled :)
